### PR TITLE
SONAR-6730 Create an API for plugins to be able compute new measures

### PR DESCRIPTION
**Pull request by julienlancelot**

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/computation/measure/api/MeasureComputerImpl.java [hunk 1]
```diff
@@ -88,41 +91,42 @@ public String toString() {
...
     }
 
-    private static void checkImplementation(Implementation impl) {
-      checkArgument(impl != null, "The implementation is missing");
+    private static String[] validateOutputMetricKeys(@Nullable String[] outputMetrics) {
+      checkArgument(outputMetrics != null && outputMetrics.length > 0, "At least one output metric must be defined");
+      return outputMetrics;
+    }
+
+    private Implementation validateImplementation(Implementation impl) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/computation/measure/api/MeasureComputerImpl.java]: MAJOR Make "validateImplementation" a "static" method. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2325)

File: server/sonar-server/src/main/java/org/sonar/server/computation/measure/api/MeasureComputerImpl.java [hunk 2]
```diff
@@ -21,10 +21,13 @@
 package org.sonar.server.computation.measure.api;
 
 import com.google.common.collect.ImmutableSet;
+import java.util.ArrayList;
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/computation/measure/api/MeasureComputerImpl.java]: MINOR Remove this unused import 'java.util.ArrayList'. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AUselessImportCheck)

File: sonar-plugin-api/src/main/java/org/sonar/api/ce/measure/MeasureComputer.java [hunk 3]
```diff
@@ -142,6 +144,11 @@
        */
       void addMeasure(String metric, String value);
 
+      /**
+       * Return list of issues of current component.
+       */
+      List<? extends Issue> getIssues();
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/ce/measure/MeasureComputer.java]: MAJOR Remove usage of generic wildcard type. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1452)

File: server/sonar-server/src/main/java/org/sonar/server/computation/issue/CloseIssuesOnRemovedComponentsVisitor.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+  }
+
+  private void closeIssuesForDeletedComponentUuids(Set<String> deletedComponentUuids) {
+    DiskCache<DefaultIssue>.DiskAppender cacheAppender = issueCache.newAppender();
+    try {
+      for (String deletedComponentUuid : deletedComponentUuids) {
+        List<DefaultIssue> issues = baseIssuesLoader.loadForComponentUuid(deletedComponentUuid);
+        for (DefaultIssue issue : issues) {
+          issue.setBeingClosed(true);
+          // TODO should be renamed
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/computation/issue/CloseIssuesOnRemovedComponentsVisitor.java]: INFO Complete the task associated to this TODO comment. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1135)

File: server/sonar-server/src/main/java/org/sonar/server/computation/issue/ComponentIssuesRepositoryImpl.java [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+  @Override
+  public void setIssues(Component component, List<DefaultIssue> issues) {
+    this.issues = requireNonNull(issues, "issues cannot be null");
+    this.component = requireNonNull(component, "component cannot be null");
+  }
+
+  @Override
+  public List<DefaultIssue> getIssues(Component component) {
+    checkState(this.component != null && this.issues != null, "Issues have not been initialized");
+    checkArgument(component == this.component,
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/computation/issue/ComponentIssuesRepositoryImpl.java]: MAJOR Change this comparison to use the equals method. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1698)

File: server/sonar-server/src/main/java/org/sonar/server/computation/issue/ComponentIssuesRepositoryImpl.java [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+import java.util.List;
+import javax.annotation.CheckForNull;
+import org.sonar.core.issue.DefaultIssue;
+import org.sonar.server.computation.component.Component;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+
+public class ComponentIssuesRepositoryImpl implements MutableComponentIssuesRepository, ComponentIssuesRepository {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/computation/issue/ComponentIssuesRepositoryImpl.java]: MINOR "org.sonar.server.computation.issue.MutableComponentIssuesRepository" is a "org.sonar.server.computation.issue.ComponentIssuesRepository" so "org.sonar.server.computation.issue.ComponentIssuesRepository" can be removed from the extension list. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1939)

**Comment by SonarTech** [comment]
SonarQube analysis reported 8 issues:
* MAJOR 3 major
* MINOR 2 minor
* INFO 3 info

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* INFO [Complete the task associated to this TODO comment.](https://github.com/SonarSource/sonarqube/blob/ce2b4afd49dd89b1cd28152a18dc1a8a32c783c0/server/sonar-server/src/main/java/org/sonar/server/computation/issue/IntegrateIssuesVisitor.java#L107) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1135)
* INFO [Complete the task associated to this TODO comment.](https://github.com/SonarSource/sonarqube/blob/ce2b4afd49dd89b1cd28152a18dc1a8a32c783c0/server/sonar-server/src/main/java/org/sonar/server/computation/issue/IntegrateIssuesVisitor.java#L105) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1135)

File: sonar-plugin-api/src/main/java/org/sonar/api/ce/measure/MeasureComputer.java (hunk 3 above)
Comment (julienlancelot) [inline sonar-plugin-api/src/main/java/org/sonar/api/ce/measure/MeasureComputer.java]: I found no other to be able to directly return to list of DefaultIssue. Removing the wildcard needs to either return a list of wrapper of the DefaultIssue or to cast every DefaultIssue in Issue -> Both will require a lot of CPUs to process as the number of issues can be huge.

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues:
* MAJOR 1 major
* INFO 3 info

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* INFO [Complete the task associated to this TODO comment.](https://github.com/SonarSource/sonarqube/blob/7c2250508acc1a7d1470390ca801c0401291bce2/server/sonar-server/src/main/java/org/sonar/server/computation/issue/IntegrateIssuesVisitor.java#L105) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1135)
* INFO [Complete the task associated to this TODO comment.](https://github.com/SonarSource/sonarqube/blob/7c2250508acc1a7d1470390ca801c0401291bce2/server/sonar-server/src/main/java/org/sonar/server/computation/issue/IntegrateIssuesVisitor.java#L107) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1135)

File: sonar-plugin-api/src/main/java/org/sonar/api/measures/RangeDistributionBuilder.java [hunk 7]
```diff
@@ -41,7 +39,9 @@
  * with pre-defined ranges of complexity.</p>
  *
  * @since 1.10
+ * @deprecated since 5.2 use {@link org.sonar.api.ce.measure.RangeDistributionBuilder instead}
  */
+@Deprecated
 public class RangeDistributionBuilder implements MeasureBuilder {
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/measures/RangeDistributionBuilder.java]: INFO Do not forget to remove this deprecated code someday. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1133)

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues:
* MAJOR 1 major
* INFO 4 info

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* INFO [Complete the task associated to this TODO comment.](https://github.com/SonarSource/sonarqube/blob/87ba0b4509963a91edd30f5636f2068cea22c362/server/sonar-server/src/main/java/org/sonar/server/computation/issue/IntegrateIssuesVisitor.java#L105) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1135)
* INFO [Complete the task associated to this TODO comment.](https://github.com/SonarSource/sonarqube/blob/87ba0b4509963a91edd30f5636f2068cea22c362/server/sonar-server/src/main/java/org/sonar/server/computation/issue/IntegrateIssuesVisitor.java#L107) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1135)

File: sonar-plugin-api/src/main/java/org/sonar/api/ce/measure/MeasureComputer.java (hunk 3 above)
Comment (simonbrandhof) [inline sonar-plugin-api/src/main/java/org/sonar/api/ce/measure/MeasureComputer.java]: ```List<Issue>``` ?

Comment (julienlancelot) [inline sonar-plugin-api/src/main/java/org/sonar/api/ce/measure/MeasureComputer.java]: See my comment just before.

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues:
* MAJOR 1 major
* INFO 4 info

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* INFO [Complete the task associated to this TODO comment.](https://github.com/SonarSource/sonarqube/blob/1fe43f8154e61b3d62b95d49ca118fb1906a2b22/server/sonar-server/src/main/java/org/sonar/server/computation/issue/IntegrateIssuesVisitor.java#L107) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1135)
* INFO [Complete the task associated to this TODO comment.](https://github.com/SonarSource/sonarqube/blob/1fe43f8154e61b3d62b95d49ca118fb1906a2b22/server/sonar-server/src/main/java/org/sonar/server/computation/issue/IntegrateIssuesVisitor.java#L105) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1135)

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues:
* MAJOR 1 major
* INFO 4 info

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* INFO [Complete the task associated to this TODO comment.](https://github.com/SonarSource/sonarqube/blob/45b6fbcffb18fe98f1d17d435b97200a61496024/server/sonar-server/src/main/java/org/sonar/server/computation/issue/IntegrateIssuesVisitor.java#L107) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1135)
* INFO [Complete the task associated to this TODO comment.](https://github.com/SonarSource/sonarqube/blob/45b6fbcffb18fe98f1d17d435b97200a61496024/server/sonar-server/src/main/java/org/sonar/server/computation/issue/IntegrateIssuesVisitor.java#L105) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1135)

File: server/sonar-server/src/main/java/org/sonar/server/computation/issue/CloseIssuesOnRemovedComponentsVisitor.java [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+ * Close issues on removed components
+ */
+public class CloseIssuesOnRemovedComponentsVisitor extends TypeAwareVisitorAdapter {
+
+  private final BaseIssuesLoader baseIssuesLoader;
+  private final ComponentsWithUnprocessedIssues componentsWithUnprocessedIssues;
+  private final IssueCache issueCache;
+  private final IssueLifecycle issueLifecycle;
+
+  public CloseIssuesOnRemovedComponentsVisitor(BaseIssuesLoader baseIssuesLoader, ComponentsWithUnprocessedIssues componentsWithUnprocessedIssues, IssueCache issueCache, IssueLifecycle issueLifecycle) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/computation/issue/CloseIssuesOnRemovedComponentsVisitor.java]: MINOR Split this 202 characters long line (which is greater than 180 authorized). [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00103)

**Comment by SonarTech** [comment]
SonarQube analysis reported 6 issues:
* MAJOR 1 major
* MINOR 1 minor
* INFO 4 info

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* INFO [Complete the task associated to this TODO comment.](https://github.com/SonarSource/sonarqube/blob/c96843eab848209960efb21ddebe5afe80cb2fd8/server/sonar-server/src/main/java/org/sonar/server/computation/issue/IntegrateIssuesVisitor.java#L107) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1135)
* INFO [Complete the task associated to this TODO comment.](https://github.com/SonarSource/sonarqube/blob/c96843eab848209960efb21ddebe5afe80cb2fd8/server/sonar-server/src/main/java/org/sonar/server/computation/issue/IntegrateIssuesVisitor.java#L105) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1135)

**Comment by SonarTech** [comment]
SonarQube analysis reported 6 issues:
* MAJOR 1 major
* MINOR 1 minor
* INFO 4 info

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* INFO [Complete the task associated to this TODO comment.](https://github.com/SonarSource/sonarqube/blob/c77463262fd6506e24d5808704cb8e19926f9c29/server/sonar-server/src/main/java/org/sonar/server/computation/issue/IntegrateIssuesVisitor.java#L105) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1135)
* INFO [Complete the task associated to this TODO comment.](https://github.com/SonarSource/sonarqube/blob/c77463262fd6506e24d5808704cb8e19926f9c29/server/sonar-server/src/main/java/org/sonar/server/computation/issue/IntegrateIssuesVisitor.java#L107) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1135)

**Comment by SonarTech** [comment]
SonarQube analysis reported 6 issues:
* MAJOR 1 major
* MINOR 1 minor
* INFO 4 info

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* INFO [Complete the task associated to this TODO comment.](https://github.com/SonarSource/sonarqube/blob/e39106a0e889a6ed9c46f18d56d2c401c00b9019/server/sonar-server/src/main/java/org/sonar/server/computation/issue/IntegrateIssuesVisitor.java#L105) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1135)
* INFO [Complete the task associated to this TODO comment.](https://github.com/SonarSource/sonarqube/blob/e39106a0e889a6ed9c46f18d56d2c401c00b9019/server/sonar-server/src/main/java/org/sonar/server/computation/issue/IntegrateIssuesVisitor.java#L107) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1135)

**Comment by SonarTech** [comment]
SonarQube analysis reported 6 issues:
* MAJOR 1 major
* MINOR 1 minor
* INFO 4 info

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* INFO [Complete the task associated to this TODO comment.](https://github.com/SonarSource/sonarqube/blob/674fb0012f9f5d4de75c6ffcd346b65a591f1681/server/sonar-server/src/main/java/org/sonar/server/computation/issue/IntegrateIssuesVisitor.java#L105) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1135)
* INFO [Complete the task associated to this TODO comment.](https://github.com/SonarSource/sonarqube/blob/674fb0012f9f5d4de75c6ffcd346b65a591f1681/server/sonar-server/src/main/java/org/sonar/server/computation/issue/IntegrateIssuesVisitor.java#L107) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1135)

**Pull request merged**